mono/metadata/threads.c

```diff
@@ -907,14 +870,11 @@ mono_thread_detach_internal (MonoInternalThread *thread)
 	 * This can happen only during shutdown.
 	 * The shutting_down flag is not always set, so we can't assert on it.
 	 */
-	if (thread->synch_cs)
-		LOCK_THREAD (thread);
-
-	thread->state |= ThreadState_Stopped;
-	thread->state &= ~ThreadState_Background;
-
-	if (thread->synch_cs)
+	if (TRY_LOCK_THREAD (thread)) {
```

Comment:
This looks wrong, the thread state must be always updated.

Reply:
I was kinda wondering about try vs. assert also, in my attempt I asserted all and was going to see where it failed.